apps/src/aichat/redux/utils.ts

```diff
@@ -0,0 +1,80 @@
+import moment from 'moment';
+
+import {getTypedKeys} from '@cdo/apps/types/utils';
+
+import {
+  AiCustomizations,
+  FieldVisibilities,
+  ModelCardInfo,
+  Visibility,
+} from '../types';
+
+// This variable keeps track of the most recent message ID so that we can
+// assign a unique message id in increasing sequence to a new message.
+let latestMessageId = 0;
+export const getNewMessageId = () => {
+  latestMessageId += 1;
+  return latestMessageId;
+};
+
+export const getCurrentTimestamp = () =>
+  moment(Date.now()).format('YYYY-MM-DD HH:mm');
+export const getCurrentTime = () => moment(Date.now()).format('LT');
+
+const haveDifferentValues = (
+  value1: AiCustomizations[keyof AiCustomizations],
+  value2: AiCustomizations[keyof AiCustomizations]
+): boolean => {
+  if (typeof value1 === 'object' && typeof value2 === 'object') {
+    return JSON.stringify(value1) !== JSON.stringify(value2);
+  }
+
+  return value1 !== value2;
+};
+
+// Used to decide which model customizations have changed
+// between the previous save and the current one,
+// such that we can display a notification for each to users.
+export const findChangedProperties = (
+  previous: AiCustomizations | undefined,
+  next: AiCustomizations
+) => {
+  if (!previous) {
+    return Object.keys(next);
+  }
+
+  const changedProperties: string[] = [];
+  Object.keys(next).forEach(key => {
+    const typedKey = key as keyof AiCustomizations;
+    if (haveDifferentValues(previous[typedKey], next[typedKey])) {
+      changedProperties.push(key);
+    }
+  });
+
+  return changedProperties;
+};
+
+// Used to decide whether to unpublish a project based on whether
+// it has its model card filled out or not.
+export const hasFilledOutModelCard = (modelCardInfo: ModelCardInfo) => {
+  for (const key of getTypedKeys(modelCardInfo)) {
+    if (key === 'isPublished') {
+      continue;
+    } else if (key === 'exampleTopics') {
+      if (
+        !modelCardInfo['exampleTopics'].filter(topic => topic.length).length
+      ) {
+        return false;
+      }
+    } else if (!modelCardInfo[key].length) {
+      return false;
+    }
+  }
+
+  return true;
```

Comment:
I think this could be smooshed down a lot, though I'm not convinced it's more legible. 

Assuming I got all my logic right...

```
// note that we return the negation - so if we find a key === 'exampleTopics' AND the modelCardInfo.exampleTopics
// has a length, OR we have no length on the modelCardInfo[key], then we return a true value (result of the `some`)
// but we want to return a false in that case, so we negate it. If we find nothing, then the `some` is false so 
// we negate it and return `true`
return !getTypedKeys(modelCardInfo).some( key => 
  (key === 'exampleTopics' && !modelCardInfo.exampleTopics.filter(topic => topic.length).length)
  || !modelCardInfo[key].length
```

Reply:
ugh yeah you're right this isn't pretty...I think your suggestion doesn't cover the `isPublished` case, which is a boolean. It'd have to be:

```
return !getTypedKeys(modelCardInfo).some( key => 
  (key === 'exampleTopics' && !modelCardInfo.exampleTopics.filter(topic => topic.length).length)
  || (key !== 'isPublished' && !modelCardInfo[key].length)
```

The various negations/logic combinations you have to do in your head for this to work feels harder to understand than the quite verbose version we have currently IMO...what do you think?

Reply:
FWIW, I think that the current implementation may be easier to understand for a future reader.